proto/v1beta1/intoto.proto

```diff
@@ -0,0 +1,145 @@
+  //   }
+  // }
+  map<string, google.protobuf.Struct> materials = 2 [json_name="materials"];
+
+  // Products are the supply chain artifacts generated as a result of the step.
+  // The structure is identical to that of materials.
+  map<string, google.protobuf.Struct> products = 3 [json_name="products"];
+
+  // By Products are made up of the data that is generated as part of a software
+  // supply chain step, but are not the actual result of the step.
+  ByProductObject byproducts = 4 [json_name="byproducts"];
+
+  // This is a field that can be used to capture information about the
+  // environment. It is recommended for this field to contain information that
+  // details environment variables, filesystem information, and the present
+  // working directory. The structure of this field is:
+  // "environment": {
+  //   "variables": "<ENV>",
+  //   "filesystem": "<FS>",
+  //   "workdir": "<CWD>"
+  // }
+  map<string, google.protobuf.Struct> environment = 5 [json_name="environment"];
+}
+
+// This corresponds to a full signed attestation - it is made up of one or more
+// signatures and the core link attestation itself. This is used for occurrences
+// of a Grafeas in-toto note.
+message Details {
+  repeated Signature signatures = 1 [json_name="signatures"];
+  Link link = 2 [json_name="signed"];
+}
+
+// This contains the private and public portions of a key. The private portion
+// must be empty for a public key.
+message KeyValue {
+  string private = 1;
+  string public = 2;
+}
+
+// This defines the format used to record keys used in the software supply
+// chain. An in-toto link is attested by a functionary using one or more keys
+// defined in the in-toto layout.
+message FunctionaryKey {
+  // keyid is an identifier for the signing key.
+  string keyid = 1;
+
+  // This field identifies the hash algorithms used in the particular key.
+  repeated string keyid_hash_algorithms = 2;
+
+  // This field identifies the specific signing method. Eg: RSASSA-PSS, ed25519,
+  // and ecdsa.
+  string keytype = 3;
+
+  // This field contains the actual private and public portions of the key.
+  KeyValue keyval = 4;
+
+  // This field contains the overall scheme used for the creation of the signing
+  // key. Eg: rsassa-pss-sha256
+  string scheme = 5;
+}
+
+// This contains the core fields corresponding to a software supply chain step.
+// This information goes into a Grafeas note.
+message InToto {
+  // This field identifies the name of the step in the supply chain.
+  string step_name = 1;
+
+  // This field contains the public keys that can be used to verify the
+  // signatures on the step metadata.
+  repeated FunctionaryKey signing_keys = 2;
+
+  // The following fields contain in-toto artifact rules identifying the
+  // artifacts that enter this supply chain step, and exit the supply chain
+  // step, i.e. materials and products of the step.
+  repeated google.protobuf.ListValue expected_materials = 3;
```

Comment:
you're already using `repeated` here, no need for `ListValue` IIUC

Reply:
These fields are set up as two-dimensional arrays of strings, this was the only way I could find to define in protobuf. Is there a better way? :)

A snippet from one of our demos:
```
   "expected_materials": [
     [
      "MATCH",
      "demo-project.tar.gz",
      "WITH",
      "PRODUCTS",
      "FROM",
      "package"
     ],
     [
      "ALLOW",
      ".keep"
     ],
     [
      "ALLOW",
      "alice.pub"
     ],
     [
      "ALLOW",
      "root.layout"
     ],
     [
      "DISALLOW",
      "*"
     ]
    ],
```

Reply:
Gotcha, so here you'll want to have a `repeated Material expected_materials = 3;` field, and define a `Material` proto message separately. If there's no structure to `Material` beyond having a list of strings, then you'd just have 1 field in the proto, `repeated string ... = 1;`. If you do know the structure ahead of time, then you'd define it in the `Material` proto itself. Does this make sense?

Reply:
I tried that before settling for `google.protobuf.ListValue` and the issue I ran into IIRC is that it expected a list of an object (which could handle repeated strings for the given key).

```
[
  {
    <key as per the proto definition>: [<list of strings>]
  }
]
```

I can double check this though.

Reply:
Resolving this as we're discussing this in a separate thread below.